### ISSUE TITLE
GUI: Spectra are not updating

### ISSUE THREAD

**Issue by John-Holt-Tessella** [ISISComputingGroup/IBEX#1968]
During the last sprint instrument scientists (Winifred and Ian) reported that the spectra would freeze on them. Both as a group of spectra and individually. Switching to a different spectra and switching back did not stop this freeze. 

I suggest pointing the spectra at a local PV and try different condition. Like disconnected. No data, 1 point of data (in x and y). Maybe even update values quickly so it reads it between updating values. If this finds nothing try looking at underlying PV and see if it is this that is not updating correctly.

On completion inform Winifred and Ian.

**Comment by AdrianPotter** [comment]
The problem appears to be in the PV manager. I can reproduce the issue in part by:

1. Switch the spectra over to use SIMPLE IoC rather than DAE in DaeObservables.java
2. Set up some appropriate PVs in SIMPLE of the form SIMPLE:SPEC:n:m:X. Use SIMPLE:LONG for the data length
3. I used genie_python to script feeding some random variables into the data e.g.

```
import epics
import numpy

pv = epics.PV("...")
while True:
    pv.put(numpy.random.rand(10))
```

4. Restart the IoC while it's doing that. After 1 or two tries, the DAE spectra will stop updating in the GUI

I tracked this down and it appears that no new values are being detected by the DaeObservables observables. This is likely an issue in PVManager. I've yet to find a way to correct the issue. Notably in the setup above, changing spectra number triggers the `update()` method in ObservedSpectra so that the DAE observable for the spectra is refreshed and updates resume. I've not yet found a way to leverage that in working around the bug.

**Comment by AdrianPotter** [comment]
Having discussed with @nickbattam, there is a reasonable chance that this problem may disappear if we update to the latest version of CSStudio 3.3, along with CAJ and JCA. Upgrading to CSStudio 4 would be ideal but this obviously has a much bigger overhead.

My view is that upgrading to CSStudio 3.3 is outside the scope of this ticket. I tried just grabbing the latest version online but we have a slightly customised version of CAJ.

- The CSS studio builds in Jenkins are currently disabled
- They are based on the old 3.3 version of the source code
- That source code currently resides in SVN

Sorting all of that out will require work beyond this ticket. I will discuss at the retrospective this afternoon and create a ticket if necessary.

I don't see a stopgap measure being available in the mean time; I can't see a way of spurring the DAE observables back into life. Creating new observables isn't a viable option since it would require tracking down all the observers and asking them to re-register. Nor would a top-down approach of just creating a new DAE model because there are too many observers outside of the plugin that would not know of the change. I think that any workaround is going to be sufficiently complex and messy as to be not worth the effort compared to upgrading CSStudio.

Notably the risk is that we upgrade to CSStudio 3.3.11 and it doesn't fix the problem. We haven't confirmed that there are issues fixed between our version of CSStudio and 3.3.11 that may resolve the problem:

https://github.com/ControlSystemStudio/cs-studio/issues/216

but the only way to find out will be to try. Of course, upgrading may also have a beneficial impact on other parts of the system, and it would be worthwhile taking back control over our own build system as far as CSStudio is concerned. Using a 2015 version that we have no way of updating or reproducing isn't a good long term solution.

**Comment by AdrianPotter** [comment]
I've created a branch on the GUI:

https://github.com/ISISComputingGroup/ibex_gui/tree/Ticket1968_Spectra_not_updating

It adds a "Reset server connection" to the help menu. Underneath, it just switches instrument to the current instrument. This has the effect of resetting all switchable PV observables and has been demonstrated to restore the described DAE behaviour along with several issues regarding other PV-dependent properties (blocks, OPIs etc.) not updating.

This is a bit (a lot) of a fudge, ideally it shouldn't be needed. I've left the branch open but first it would be good to try and fix Ticket #1229 to see if it resolves the issue.

If this issue hasn't been resolved by with a sprint, then I'd recommend adding this so it is available for the next release. It can always be removed easily at a later date should it become unnecessary.

**Comment by Tom-Willemsen** [comment]
#1229 has been merged now. Marking as ready rather than impeded.

**Comment by AdrianPotter** [comment]
Thanks. I'll investigate whether the new version of CSStudio has changed the behaviour

**Comment by AdrianPotter** [comment]
I've updated my version of CCStudio to the new version but it hasn't had any effect on this problem :(.

As suggested I'll create a PR for the server connection reset button in the GUI.

**Comment by mattclarke** [comment]
It is suspected that this might be an indicator of a greater problem relating to gateways and beacons.
I am setting up NDXICAP to leave a test running to see if we can reproduce the freeze.

**Comment by mattclarke** [comment]
The beacons and gateways issue is now in its own ticket: https://github.com/ISISComputingGroup/IBEX/issues/2163

**Comment by AdrianPotter** [comment]
The issue couldn't be reproduced on NDXICAP even after being left watching spectra for a protracted period of time. We need to try another line of inquiry since this is a high priority fix.

One hypothesis is that PV manager's buffer is being overloaded. This might happen if it retains a connection to many spectra PV. The issue may only be triggered when spectra have been switched many times.

1. Most tests we've done use a limited number of simulated spectra. Can we extend the simulated DAE to provide more non-trivial (i.e. evolving) spectra?
2. Can we look at one of the instruments and monitor the network traffic as we change spectra. Is there a noticeable increase? Does it max out?
3. Can we observe the instruments triggering the error? Is there a particular workflow that can reliably reproduce it?

**Comment by AdrianPotter** [comment]
If an instrument scientist reports seeing this issue, please try a `caget` on `IN:[INST]:DAE:SPEC:n:m:a` where `n` is the spectrum number (try some small numbers), `m` is the period number (1 should be fine) and `a` is the axis (try `X` and `Y`). For instance, on Vesuvio please try:

`IN:VESUVIO:DAE:SPEC:1:1:X`
`IN:VESUVIO:DAE:SPEC:1:1:Y`
`IN:VESUVIO:DAE:SPEC:2:1:X`
`IN:VESUVIO:DAE:SPEC:2:1:Y`

**Comment by AdrianPotter** [comment]
Based on investigation so far, I believe the issue is in the ISISDAE IOC.

1. **The GUI only updates when information is received**: When looking at the DAE spectra tab, the symptoms reported by the instrument scientists is that the GUI freezes. This is symptomatic of no new data or connection timeouts from the camonitor on the spectra PVs
1. **`caget` frequently times out locally**: When asking for a new spectra using `caget` from the ISISDAE, the data is often slow to return. During testing on Alf, this was anything up to a minute. This explains the behaviour in point 1. If scientists change to a spectra, it may not connect initially. The monitor interval on the spectra PV would be extended after the initial failed connection. This would cause the GUI to look 'frozen'. When the user restarts the GUI, the PV has frequently become available so then appears on reconnection. Notably at the start of cycle on Vesuvio, there were instances where the PV wasn't available even after restarting the GUI
1. **The delay on spectra PV becoming available is much longer over the network**: When doing a `caget` over the network rather than on localhost, the response time can be of the order of several minutes. This was tested between NDW1695 (my machine) and Alf. This doesn't necessarily relate to the problem directly, it may be an issue with the gateway.
1. **The problem hasn't been reproduced on a simulated DAE**: I've tried asking for lots of spectra and bombarding the DAE IOC with lots of requests but none of them have shown the same behaviour as on instruments.

Having attached to the Alf ISISDAE IOC remotely using Visual Studio 2010, there aren't obvious points in the code that are causing the bottleneck. Once the spectra PV has been requested, the methods for getting the data clear very quickly. However, there is frequently a significant delay in the methods being triggered, and often a delay after it has been completed before it becomes available in the PV. On Alf, several messages were reported about excessive scan times. The default scan rate for created PvInfo objects is 0.5 seconds. This leaves several lines of enquiry:

1. **Can the number of monitored PVs be reduced?** Once a PV has been requested, the IOC continues to monitor it indefinitely. If the IOC is being overloaded, this could create problems for long experiment sessions. We may benefit by dropping PVs that are no longer being requested.
1. **Can the GUI be clearer about when awaiting PVs** The GUI at the moment appears frozen because the old spectra is being left in place. Even if the new spectra isn't available immediately, the GUI should be clear that it is waiting on background operations to complete
1. **Does adjusting the scan rate help solve the problem?** Can we drop the scan rate from 0.5 seconds to help improve performance? This is probably only a short-term measure, it doesn't fix the underlying problem
1. **Is there a bottleneck in the ISISDAE?** Visual Studio 2017 Professional comes with profiling tools. These should be run on one of the instruments that is experiencing issues. Alf is currently in use, there is a window of opportunity to run some testing between 10th and 12th May. I'll try running it then to see if there's any evidence of a bottleneck in the IOC.

**Comment by AdrianPotter** [comment]
Setting to impeded. I'll check points 1, 3 and 4 when Alf isn't in use

**Comment by AdrianPotter** [comment]
I had a look at this on Friday. On point 1, the PVs are slow to respond immediately after that start of the ISISDAE IOC so I don't think this will have any effect. On point 3, I tried reducing the scan rate of the PVInfo objects to 2 seconds rather than 0.5 which had no visible effect on the functionality. 

I've been unable to test point 4 because although my copy of VS 2017 has profiling tools, it doesn't seem to have tools for profiling on remote machines. @FreddieAkeroyd, did you say you had them? If so, might you be able to drop them on Alf so we can take a look inside the DAE IOC at some point?

**Comment by FreddieAkeroyd** [comment]
I think I said my visual studio 2010 had profiling tools as it is "ultimate" rather than professional, I've only ever done remote debugging rather than remote profiling. I thought you were going to use the standalone profiler on ALF?

**Comment by AdrianPotter** [comment]
@Tom-Willemsen mentioned that the DAE executable may return slowly if it's asked for very large integrals in the case of the run diagnostics 2 page. It might be worth taking a look at the VI and comparing its calls to the DAE exec compared to the DAE.

**Comment by AdrianPotter** [comment]
Spectra (and other DAE) PVs operate in two modes:

- A slow loop when they're not being monitored (currently 5s)
- A fast loop when they're being monitored (0.5s)

In the GUI, spectra were created at startup and then left in the fast monitoring loop. All inspected spectra remained on this loop indefinitely, even if the selected spectra is changed in the DAE perspective. This is compounded by use of the incremental spectra buttons in the DAE perspective which result in all intermediate spectra being monitored between the start and ending values. This puts a large and unnecessary load on the DAE. In the upcoming PRs, I've made the following changes:

- Decreased the scan rate on the slow loop to 30s. This significantly decreases the load from unmonitored PVs. I haven't shut them off entirely because it may be beneficial to have a not too out-of-date value available at short notice.
- Only created new spectra observables once the incremental (up/down) spectra buttons have been left idle for 1 second. This means there is less likelihood of creating lots of intermediate spectra.
- I've actively closed the spectra observables when we change spectra. This disconnects the underlying PV observable. Once the object is garbage collected, the unwatched spectra now move from the fast to the slow loop

Owing to the difficulty in reproducing the error, it is difficult to definitively say whether this will solve the problem. I've tested the new executable on Alf and it responds as expected. However, it is clear that these changes will significantly reduce the load on the ISISDAE IOC from the executable which should improve performance.
